src/main/java/com/epam/esm/dao/GiftCertificateDao.java

```diff
@@ -0,0 +1,70 @@
+package com.epam.esm.dao;
+
+import com.epam.esm.entity.GiftCertificate;
+import com.epam.esm.entity.SqlQueries;
+import com.epam.esm.mapper.GiftCertificateMapper;
+import com.epam.esm.utils.QueryConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Component;
+import java.sql.PreparedStatement;
+import java.sql.Statement;
+import java.util.List;
+import java.util.Optional;
+
+@Component
+public class GiftCertificateDao implements EntityDao<GiftCertificate> {
+
+    private static final String CREATE_GIFT_CERTIFICATE = "INSERT INTO gift_certificate (name, description, price, duration, create_date, last_update_date) VALUES (?, ?, ?, ?, ?, ?)";
+
+    private final JdbcTemplate template;
+    private final GiftCertificateMapper mapper;
+    private final QueryConstructor constructor;
+
+    @Autowired
+    public GiftCertificateDao(JdbcTemplate template, GiftCertificateMapper mapper, QueryConstructor constructor) {
+        this.template = template;
+        this.mapper = mapper;
+        this.constructor = constructor;
+    }
+
+    @Override
+    public void deleteById(long id) {
+        template.update(SqlQueries.DELETE_CERTIFICATE, id);
+    }
+
+    @Override
+    public long create(GiftCertificate giftCertificate) {
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+        template.update(connection -> {
+            PreparedStatement statement = connection.prepareStatement(SqlQueries.CREATE_CERTIFICATE, Statement.RETURN_GENERATED_KEYS);
+            statement.setString(1, giftCertificate.getName());
+            statement.setString(2, giftCertificate.getDescription());
+            statement.setString(3, giftCertificate.getPrice().toString());
+            statement.setString(4, String.valueOf(giftCertificate.getDuration().toDays()));
+            statement.setString(5, giftCertificate.getCreationDate());
+            statement.setString(6, giftCertificate.getLastUpdateDate());
+            return statement;
+        }, keyHolder);
+        return keyHolder.getKey().longValue();
+    }
+
+    public List<GiftCertificate> getGiftCertificates(String tagName, String giftCertificateName, String description, String sortByName, String sortByDate) {
+        String query = constructor.constructQuery(tagName, giftCertificateName, description, sortByName, sortByDate);
+        return template.query(query, mapper);
+    }
+
+    @Override
+    public Optional<GiftCertificate> findById(long id) {
+        return template.query(SqlQueries.FIND_CERTIFICATE_BY_ID, mapper, new Object[]{id}).stream().findAny();
```

Comment:
Sonarlint says that we can catch NullPointer here:
![image](https://user-images.githubusercontent.com/17936204/131492559-29f6cb5f-efd0-48f1-a3af-5ce46c2f1121.png)
And I'm not sure why do we need to apply these methods .stream().findAny() here?

Reply:
We need the methods because it returns a List, while we need Optional to be returned

Reply:
and why cannot we have null check?